url7606378bcc_ljy2010a_protobuf2_tgz-pJ8-com_game_gameObj_bufBuilder_ReflectBuilderJ8/src/com/game/gameObj/bufBuilder/BufBuilderMake.java

```diff
@@ -0,0 +1,140 @@
+package com.game.gameObj.bufBuilder;
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.PrintWriter;
+
+public class BufBuilderMake {
+
+	public static String[] typeNames = {"Soldier","WarShip"};
+	
+	/**
+	 * @param args
+	 * @throws ClassNotFoundException 
+	 * @throws IOException 
+	 */
+	public static String FilePath = "src/com/game/gameObj/bufBuilder/";
+	
+	public static String GameObjId  = "GameObjId";
+	
+	public static String GameObjType  = "GameObjType";
+	
+	public static String Coor = "Coor";
+	
+	
+	public static void main(String[] args) throws ClassNotFoundException, IOException {
+		// TODO Auto-generated method stub
+		for (int i = 0 ;i<typeNames.length;i++){
+			PrintWriter pw = new PrintWriter( new FileWriter( FilePath+ typeNames[i] + "Builder.java" ) );
+	        pw.print(makeBuilder(typeNames[i]));
+	        pw.close();
+		}
+	}
+	
+	public static String makeBuilder(String arg_TypeName) throws ClassNotFoundException{
+		String TypeName = arg_TypeName;
+		String AttrName = arg_TypeName+"Attr";
+		String Head = "";
+		
+		Head +=  "package com.game.gameObj.bufBuilder; \n\n";
+		
+		Head +=  "import com.game.gameObj."+TypeName+";\n";
+		
+		Head +=  "import com.game.protobuf.GameProto;\n";
+		
+		Head +=  "import com.game.protobuf.ModelAttr;\n";
+		
+		Head +=  "import com.game.protobuf.ModelAttr."+AttrName+";\n\n";  
+		
+		Head +=  "public class "+TypeName+"Builder{ \n\n";
+		
+		Head +=  BuildFunc(TypeName,true);
+		
+		Head +=  BuildFunc(TypeName,false);
+        
+        Head +="}\n";
+        
+		return Head;		
+	}
+	
+	
+	public static String BuildFunc(String arg_TypeName,boolean isAllFunc) throws ClassNotFoundException{
+		String TypeName = arg_TypeName;
+		String AttrName = arg_TypeName+"Attr";
+		String Head = "";
+				
+		if(!isAllFunc){
+			Head +=  "public static GameProto.GameAction getGameAction("+TypeName+" arg_Origin,boolean isSyncCoor ,String ActionType,"+AttrName+"... Attrs){\n\n";	
+		}else{
+			Head +=  "public static GameProto.GameAction getGameAction(String ActionType,"+TypeName+" arg_Origin){\n\n"; 
+		}				 
+
+		Head +=  "\t\tGameProto.GameObject.Builder gameGameObjectBuilder = GameProto.GameObject.newBuilder();\n\n";
+		
+        Head +=  "\t\tModelAttr."+TypeName+".Builder Builder = ModelAttr."+TypeName+".newBuilder();\n\n";
+        
+        if(!isAllFunc){
+        	Head +=  "\t\tfor (int i = 0; i<Attrs.length ; i++ ){\n\n";
+        }
+        @SuppressWarnings("rawtypes")
+		Class c = Class.forName("com.game.gameObj."+TypeName); 
+       
+        for(int i = 0 ; i < c.getFields().length; i++){
+        	//TODO: GameObjId , gameObjType is not use here
+        	
+//        	System.out.println(c.getFields()[i].getName());
+//        	System.out.println(c.getFields()[i].getType().getName());
+        	String FieldsName = c.getFields()[i].getName();
+        	
+        	if(FieldsName == GameObjId || FieldsName == GameObjType || FieldsName == Coor){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.